File: maxLength.java
class Solution {
    public int findMaxLength(int[] nums) {
       int t=0;
       int l=0;
       HashMap<Integer,Integer> map = new HashMap<>();
       map.put(0,-1);
       for(int i=0; i<nums.length;i++){
          if(nums[i]==0){
            nums[i]=t-1; t=t-1;
          }
            else{
             nums[i]=t+1;
            t=t+1;
            }
            
            if(!map.containsKey(nums[i]))
             map.put(nums[i],i);
             if(map.containsKey(nums[i])){
                if(l<i-map.get(nums[i]))
                   l=i-map.get(nums[i]);
             }
       } 
       return l;
    }
}

File: max_length.py
class Solution:
    def findMaxLength(self, nums: List[int]) -> int:

        nums = [-1 if num == 0 else num for num in nums]  

        sum_dict = {0: -1} 
        previous_sum = 0
        max_length = 0

        for i in range(len(nums)):
            previous_sum += nums[i]

            if previous_sum in sum_dict:
                max_length = max(max_length, i - sum_dict[previous_sum])
            else:
                sum_dict[previous_sum] = i 

        return max_length
